Adjustments for COT compatibility

asumaran opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Part of #2407

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

Minor adjustments suggested by the [COT upgrade recipe book](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#apis-for-gettingsetting-posts-and-postmeta).

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

---
**To test this PR you need to install the specific WooCommerce testing version that introduces the Custom Order Table implementation. You can [download it from here](https://github.com/woocommerce/woocommerce/releases/tag/feature-custom-order-table). Then follow [this steps to enable COT](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#easy-way).** There's also another [option using GIT](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#using-gh).

Then go to "WooCommerce > Settings > Advanced > Custom data stores" and make sure "Use the WooCommerce order tables" is selected and "Keep the posts table and the orders tables synchronized" is unchecked.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/1025173/187348641-6fcb67b3-642c-4235-b61f-8fd095fe8571.png">


---

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->


### Changes in the `get_instructions` in the `WC_Gateway_Stripe_Multibanco` class can be tested with the following steps:

- Switch currency to Euro.
- Enable Multibanco.
- Checkout using Multibanco.
- Verify the Multibanco details are visible on the "Order received" page and the email sent to the buyer.

### Fix in the `is_retention_expired` in the `WC_Stripe_Privacy` class can be tested with:

(Note that this fails on PHP 8 only. On PHP 7 the undefined array item return `null` and the multiplication will pass. On PHP 8 though it will end up throwing a fatal error because the undefined `$retention['number']` will return `""`.)

```
2022-08-30T03:44:08+00:00 CRITICAL Uncaught TypeError: Unsupported operand types: string * int in /Users/asumaran/www/woocommerce-gateway-stripe/includes/admin/class-wc-stripe-privacy.php:413
```

- You'll need to generate an order by buying a product or you can use existing orders.
- Go to Tools > Erase Personal Data
- Enter an existing customer email (check the email is not used by an already registered user in your WP install) then click the "Send Request" button.
- Hover the table row with the customer email then click "Force erase personal data"
- Make sure the request finishes without errors.

### Removed method from the `WC_Helper_Product` class can be tested with:

- `npm run up`
- `npm run test:php`
- Make sure the tests run without errors.

### Fix in the `get_order_by_source_id` method:

- Switch currency to Euro.
- Enable Multibanco.
- Create a simple product.
- Add the simple product to the cart
- Checkout using Multibanco.
- Make sure the order status is set to "Processing"
- Make sure there's no "Could not find order via source ID: seti_xyz..." message in the WC Stripe logs.

### Fix in the `get_order_by_charge_id` method:

- Create a simple product.
- Add the simple product to the cart
- Checkout using the card `4000000000000259` to [simulate disputed transaction](https://stripe.com/docs/testing#disputes).
- Make sure the order status is set to "On hold".
- Make sure there's no "Could not find order via charge ID: xyz" message in the WC Stripe logs.

### Fix in the `get_order_by_intent_id` method:

- Create a simple product.
- Add the simple product to the cart.
- Checkout using simple test card `4242424242424242`.
- Make sure the order is created successfully.
- Make sure there's no "Could not find order via intent ID: pi_xyz" message in the WC Stripe logs.

### Fix in the `get_order_by_setup_intent_id` method:

Given that WC Subscriptions is not currently supported by Custom Order Tables there's no way to test this while COT is enabled.

One way to check the fix works is to **switch back to WP post tables**, edit the `includes/class-wc-stripe-helper.php` file and negate the condition in the line 475 so the path of the fix is executed.

```php
if ( ! OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

These steps are partly based on the testing steps from [the original PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1024) that introduced the `get_order_by_setup_intent_id` method.

- Create a subscription with free trial.
- Sign up for that subscription, go to the checkout.
- Use `4000002500003155` credit card, see that the order is created normally and the subscription too.
- Make sure there's no "Could not find order via setup intent ID: xyz" message in the WC Stripe logs.

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

asumaran committed:
Fixes math operation on empty string

`wc_parse_relative_date_option` will return an array. For that reason, it will cause to skip the validation because it will never be empty.

Later below, `$retention[‘number’]` is attempted unsuccessfully because `$retention[‘number’]` will be an empty string when the “Retain Stripe Data” option is not defined, causing an error and stopping the data removal.

asumaran committed:
User Order method to update meta data

asumaran committed:
Remove unused method

asumaran committed:
Get order meta by using the `get_meta` method

reykjalin commented:
After clicking **Force erase personal data** I get this error:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/13835680/187823050-e51efcea-7405-4b3b-a9ad-7a77b9a3ae9b.png">

Is this expected? FWIW this is a brand new order and seems to be hitting a condition before the change to the retention function was made 🤔

asumaran commented:
@reykjalin I think it's because of the retention duration. What if you do an early return here? https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/class-wc-stripe-privacy.php#L404 I forgot to mention that. 

```php
public function is_retention_expired( $created_date ) {
  return true;
```

If no retention duration is defined then the data is not going to be removed. That's how it was working before.

asumaran commented:
@reykjalin Also, make sure the email was not used for a registered user. It'll only work for one of them https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/class-wc-stripe-privacy.php#L79-L83

reykjalin commented:
> Also, make sure the email was not used for a registered user. It'll only work for one of them

Yep, made sure of that. I'm using an email from a guest checkout session.

> I think it's because of the retention duration. What if you do an early return here?

Still get an error:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/13835680/187825315-7bde5ed6-af60-4074-a2c5-c9fba35cf20e.png">

reykjalin commented:
Also, Multibanco orders stay on hold. Is that supposed to happen? It's like the payment is never created 🤔

asumaran commented:
@reykjalin To test the data eraser you need to use regular orders not necessarily orders that were paid using Multibanco.

Also, The idea was to test that data erasure didn't return an error. The error on the `is_retention_expired` function happens on PHP 8 only. So, if it doesn't return an error that's fine wether you are using php 7 or php 8. If the execution makes it to the line of the fix then that's correct. Doesn't matter if it ends up erasing order data or not. Those warnings you are getting are totally fine.

asumaran commented:
> Also, Multibanco orders stay on hold. Is that supposed to happen? It's like the payment is never created 🤔

Sometimes it takes a few minutes to change to "Processing" I also noticed that with Multibanco. Unless the webhook is not properly set up?

reykjalin commented:
> To test the data eraser you need to use regular orders not necessarily orders that were paid using Multibanco.

Yeah, I only used that email for card payments.

> Also, The idea was to test that data erasure didn't return an error. The error on the is_retention_expired function happens on PHP 8 only. So, if it doesn't return an error that's fine wether you are using php 7 or php 8. If the execution makes it to the line of the fix then that's correct. Doesn't matter if it ends up erasing order data or not. Those warnings you are getting are totally fine.

Ok perfect, I won't worry about it then 🙂 

> Sometimes it takes a few minutes to change to "Processing" I also noticed that with Multibanco. Unless the webhook is not properly set up?

Yeah maybe, but the payments don't even appear in the Stripe dashboard. It's getting really late now so I'll take another look tomorrow and see if I can figure out what's going wrong.

reykjalin commented:
This is the only webhook sent from Stripe, and it's definitely getting delivered 🤔 

<img width="978" alt="image" src="https://user-images.githubusercontent.com/13835680/188072079-35a1c3e8-b207-47c4-85fb-d1e528c4ee7c.png">

Apparently my store just responded to that with a `HTTP 200` and no response body. Not sure if that's expected:

<img width="800" alt="image" src="https://user-images.githubusercontent.com/13835680/188072516-03245862-07bc-4e3f-b9dc-54b827bbf4b8.png">

I'm going to dig into this further. Maybe my multibanco setup isn't right 🤔

reykjalin commented:
It seems there's a problem with order lookup by source 🤔 

<img width="574" alt="image" src="https://user-images.githubusercontent.com/13835680/188073009-3e5f28ce-d4ee-4e53-aa9e-4c50fa7addca.png">

reykjalin commented:
I found the culprit. We're still searching the old order tables (`wp_posts`) when processing webhook payments instead of the custom order tables here:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/862f5b7ea0b8b671ac910f74d45818a20f45aeac/includes/class-wc-stripe-helper.php#L373-L383

I'm not sure how to fix that, but I assume that work should be part of this PR? 🤔

asumaran commented:
@reykjalin Nice catch! I'll fix that tomorrow.

asumaran committed:
Tweaks `get_order_by_charge_id` to be compatible with Custom Order Tables

asumaran committed:
Tweaks `get_order_by_source_id` to be compatible with Custom Order Tables

Avoids using the `wp_posts` table to get the order ID, instead uses the `wc_get_orders` which is compatible with COT.

asumaran committed:
Tweaks `get_order_by_intent_id` to be compatible with Custom Order Tables

asumaran committed:
Tweaks `get_order_by_setup_intent_id` to be compatible with Custom Order Tables

asumaran committed:
Remove unused method `delete_attribute`

asumaran committed:
Add changelog entries

asumaran committed:
Check OrderUtil class exists before using it

asumaran commented on includes/class-wc-stripe-helper.php:
```diff
@@ -373,7 +375,20 @@ public static function get_webhook_url() {
 		global $wpdb;
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $source_id, '_stripe_source_id' ) );
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
`OrderUtil` it's been recently introduced ([~2 months ago](https://github.com/woocommerce/woocommerce/pull/33638)) so for older WC versions this will throw an error. I'll add a validation

asumaran replied:
done b79149003bb0318dead088f6b8dacfd6bcba6fbf

asumaran commented on tests/phpunit/helpers/class-wc-helper-product.php:
```diff
@@ -345,23 +345,6 @@ public static function create_attribute( $raw_name = 'size', $terms = [ 'small'
-	 *
-	 * @since 2.3
-	 */
-	public static function delete_attribute( $attribute_id ) {
```
Removed this because isn't used anywhere and it's querying the database directly. It seems [`wc_delete_attribute`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-attribute-functions.php#L661) should be used instead.

asumaran commented on includes/class-wc-stripe-helper.php:
```diff
@@ -373,7 +375,20 @@ public static function get_webhook_url() {
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $source_id, '_stripe_source_id' ) );
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			$orders   = wc_get_orders(
```
Worth mentioning this works for both COT and WP post tables. I wasn't sure if we should use `wc_get_orders` for both though. Maybe in a few months when COT has been highly tested we can remove the query and use `wc_get_orders` for both scenarios.

asumaran replied:
Had a brief chat with @RadoslavGeorgiev about this since I thought he introduced the SQL query on one of his [PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/826/files?file-filters%5B%5D=.php&show-viewed-files=true#diff-dbc2cfbfe4b247c27c20d86922819977adc6f7723c0b047cb9722a2f47b7ee73R440).

Here's his response when I asked why a SQL query is used instead of `wc_get_orders`.

>It was done for performance issues. The meta table on WP sites can be absolutely abused by a ton of extensions, which is why they switched to this query.

(For example, here's a PR where the SQL query was optimized even more: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/637 )

I completely agree. The purpose of COT is to have a separate table to optimize the data lookup for WooCommerce, and I think there's no reason to use a plain SQL query here. `wc_get_orders` should be enough.

reykjalin replied:
I agree with how you've approached this and that we shouldn't be messing with the non-COT query 👍

asumaran commented:
@reykjalin Fixed the issue you found and a few more. Please review 👍

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -373,7 +375,20 @@ public static function get_webhook_url() {
+					],
+				]
+			);
+			$order_id = current( $orders )->get_id();
```
`current` may return `false` so I think we need to handle that somehow.

```suggestion
			$order_id = current( $orders ) ? current( $orders )->get_id() : false;
```

Returning `false` is really optional. So long as the default value causes `empty()` to evaluate to `true` anything should be fine.

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -396,7 +411,17 @@ public static function get_order_by_charge_id( $charge_id ) {
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			$orders   = wc_get_orders(
+				[
+					'transaction_id' => $charge_id,
```
Is this supposed to be a `meta_query` like the intent and source ID lookups or does this work as intended like this? 🤔

asumaran replied:
I had the same question and used a plain SQL query initially but I was suggested to do it this way instead p1662443099366519-slack-C9QDV4TL4

reykjalin replied:
Ok, nice!

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -373,7 +375,20 @@ public static function get_webhook_url() {
 		global $wpdb;
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $source_id, '_stripe_source_id' ) );
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
The `class_exists()` call needs to use the fully qualified class name otherwise this won't work.

```suggestion
		if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -396,7 +411,17 @@ public static function get_order_by_charge_id( $charge_id ) {
 		}
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $charge_id, '_transaction_id' ) );
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
The `class_exists()` call needs to use the fully qualified class name otherwise this won't work.

```suggestion
		if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -415,7 +440,20 @@ public static function get_order_by_charge_id( $charge_id ) {
 		global $wpdb;
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $intent_id, '_stripe_intent_id' ) );
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
The `class_exists()` call needs to use the fully qualified class name otherwise this won't work.

```suggestion
		if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -434,7 +472,20 @@ public static function get_order_by_intent_id( $intent_id ) {
 		global $wpdb;
 
-		$order_id = $wpdb->get_var( $wpdb->prepare( "SELECT DISTINCT ID FROM $wpdb->posts as posts LEFT JOIN $wpdb->postmeta as meta ON posts.ID = meta.post_id WHERE meta.meta_value = %s AND meta.meta_key = %s", $intent_id, '_stripe_setup_intent' ) );
+		if ( class_exists( 'OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```
The `class_exists()` call needs to use the fully qualified class name otherwise this won't work.

```suggestion
		if ( class_exists( 'Automattic\WooCommerce\Utilities\OrderUtil' ) && OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

reykjalin reviewed (changes requested):
Tests well, but there are a couple of things that need to be fixed before we merge 🙂

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -396,7 +411,17 @@ public static function get_order_by_charge_id( $charge_id ) {
+					'limit'          => 1,
+				]
+			);
+			$order_id = current( $orders )->get_id();
```
```suggestion
			$order_id = current( $orders ) ? current( $orders )->get_id() : false;
```

asumaran replied:
done [874b8c2](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2420/commits/874b8c2ee8a21f0c4111468e2ac09e8b28fad0ec)

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -415,7 +440,20 @@ public static function get_order_by_charge_id( $charge_id ) {
+					],
+				]
+			);
+			$order_id = current( $orders )->get_id();
```
```suggestion
			$order_id = current( $orders ) ? current( $orders )->get_id() : false;
```

asumaran replied:
done [874b8c2](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2420/commits/874b8c2ee8a21f0c4111468e2ac09e8b28fad0ec)

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -434,7 +472,20 @@ public static function get_order_by_intent_id( $intent_id ) {
+					],
+				]
+			);
+			$order_id = current( $orders )->get_id();
```
```suggestion
			$order_id = current( $orders ) ? current( $orders )->get_id() : false;
```

asumaran replied:
done [874b8c2](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2420/commits/874b8c2ee8a21f0c4111468e2ac09e8b28fad0ec)

reykjalin reviewed (approved):
There is still unchecked usage of the `current` functions which I marked in comments, but otherwise this looks great and tests well so I'm approving the PR 🙂 

Let's just add checks where we use `current()` and then merge 🚀

asumaran committed:
Use full path for class existence validation

asumaran committed:
Check if current order exist
